mods/pageviews.js

```diff
@@ -0,0 +1,309 @@
+    //   dates were the part of the query that wasn't found.
+    return res;
+};
+
+/**
+ * Parameter validators
+ */
+var throwIfNeeded = function(errors) {
+    if (errors && errors.length) {
+        throw new rbUtil.HTTPError({
+            status: 400,
+            body: {
+                type: 'invalid_request',
+                detail: errors,
+            }
+        });
+    }
+};
+
+/**
+ * Cleans the project parameter so it can be passed in as either en.wikipedia.org or en.wikipedia
+ */
+var stripOrgFromProject = function(rp) {
+    rp.project = rp.project.replace(/\.org$/, '');
+};
+
+
+var validateTimestamp = function(timestamp) {
+    if (!/^[0-9]{10}$/.test(timestamp)) {
+        return false;
+    }
+
+    var year = timestamp.substring(0, 4);
+    var month = timestamp.substring(4, 6);
+    var day = timestamp.substring(6, 8);
+    var hour = timestamp.substring(8, 10);
+
+    var dt = new Date([year, month, day].join('-') + ' ' + hour + ':00:00 UTC');
+
+    return dt.toString() !== 'Invalid Date'
+        && dt.getUTCFullYear() === parseInt(year, 10)
+        && dt.getUTCMonth() === (parseInt(month, 10) - 1)
+        && dt.getUTCDate() === parseInt(day, 10)
+        && dt.getUTCHours() === parseInt(hour);
+};
+
+
+var validateStartAndEnd = function(rp) {
+    var errors = [];
+
+    stripOrgFromProject(rp);
+
+    if (!validateTimestamp(rp.start)) {
+        errors.push('start timestamp is invalid, must be a valid date in YYYYMMDDHH format');
+    }
+    if (!validateTimestamp(rp.end)) {
+        errors.push('end timestamp is invalid, must be a valid date in YYYYMMDDHH format');
+    }
+
+    if (rp.start > rp.end) {
+        errors.push('start timestamp should be before the end timestamp');
+    }
+
+    throwIfNeeded(errors);
+};
+
+var validateYearMonthDay = function(rp) {
+    var errors = [];
+
+    stripOrgFromProject(rp);
+
+    if (rp.year === 'all-years' && (rp.month !== 'all-months' || rp.day !== 'all-days')) {
+        errors.push(
+            'month must be "all-months" and day must be "all-days" when passing "all-years"'
+        );
+    }
+    if (rp.month === 'all-months' && rp.day !== 'all-days') {
+        errors.push('day must be "all-days" when passing "all-months"');
+    }
+
+    // fake a timestamp in the YYYYMMDDHH format so we can reuse the validator
+    var validDate = validateTimestamp(
+        rp.year +
+        ((rp.month === 'all-months') ? '01' : rp.month) +
+        ((rp.day === 'all-days') ? '01' : rp.day) +
+        '00'
+    );
+
+    if (rp.year !== 'all-years' && !validDate) {
+        errors.push('year must be "all-years" or a valid 4-digit year');
+    }
+
+    if (rp.month !== 'all-months' && !validDate) {
+        errors.push('month must be "all-months" or a valid 2-digit month');
+    }
+
+    if (rp.day !== 'all-days' && !validDate) {
+        errors.push('day must be "all-days" or a valid 2-digit day');
+    }
+
+    throwIfNeeded(errors);
```

Comment:
I'm not sure we need to provide such level of detail here. For 2014-01-49 all 3 errors would happen, while year and month are valid. I think just returning smth like 'Invalid timestamp' would be enough